nakama/events.go

```diff
@@ -0,0 +1,102 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/heroiclabs/nakama-common/runtime"
+)
+
+type Event struct {
+	message string
+}
+
+type connectionChannelPairs struct {
+	connection *websocket.Conn
+	channel    chan *Event
+}
+
+type EventHub struct {
+	inputConnection *websocket.Conn
+	channels        *sync.Map //map[string]chan *Event
+	didShutdown     atomic.Bool
+}
+
+func createEventHub() (*EventHub, error) {
+	url := makeWebSocketURL(eventEndpoint)
+	webSocketConnection, _, err := websocket.DefaultDialer.Dial(url, nil)
+	for err != nil {
+		if errors.Is(err, &net.DNSError{}) {
+			//sleep a little try again...
+			fmt.Println("No host found.")
+			fmt.Println(err.Error())
+			time.Sleep(2 * time.Second)
+			webSocketConnection, _, err = websocket.DefaultDialer.Dial(url, nil)
+		} else {
+			return nil, err
+		}
+	}
+	channelMap := sync.Map{}
+	res := EventHub{
+		inputConnection: webSocketConnection,
+		channels:        &channelMap,
+		didShutdown:     atomic.Bool{},
+	}
+	res.didShutdown.Store(false)
+	return &res, nil
+}
+
+func (eh *EventHub) subscribe(session string) chan *Event {
+	channel := make(chan *Event)
+	eh.channels.Store(session, channel)
+	return channel
+}
+
+func (eh *EventHub) unsubscribe(session string) {
+	eh.channels.Delete(session)
+}
+
+func (eh *EventHub) shutdown() {
+	eh.didShutdown.Store(true)
+}
+
+// dispatch continually drains eh.inputConnection (events from cardinal) and sends copies to all subscribed channels.
+// This function is meant to be called in a goroutine.
+func (eh *EventHub) dispatch(log runtime.Logger) error {
+	var err error
+	for !eh.didShutdown.Load() {
+		messageType, message, err := eh.inputConnection.ReadMessage() //will block
+		if err != nil {
+			break
+		}
+		if messageType != websocket.TextMessage {
+			break
+		}
+		eh.channels.Range(func(key any, value any) bool {
+			channel, ok := value.(chan *Event)
+			if !ok {
+				err = errors.New("not a channel")
+				return false
+			}
+			channel <- &Event{message: string(message)}
+			return true
+		})
+		if err != nil {
+			break
+		}
+	}
+	eh.didShutdown.Store(true)
```

Comment:
This looks redundant; we'll only reach here when didShutdown is found to be true on line 72

Reply:
No there are breaks in the for loop.

Reply:
good point. 👍